### PR TITLE
feat(SCT-844): Add panel approval

### PR DESCRIPTION
**Pull request by anabebs**
**What**  
Screenshot from 2021-07-19 17-00-29
Screenshot from 2021-07-19 17-00-34
Screenshot from 2021-07-19 17-00-22

Adds frontend changes required for panel approval. This includes a new panel approval widget and dialogue box. 

**Why**  
Once a manager has approved a submission they will see a prompt to select whether the submission has also had panel approval or should be returned for edits.

**Anything else?**

The back end code is work in progress.

- Anything else in this PR that isn't covered by the what/why?
Tests will need to be written

### CODE REVIEW

File: components/ApprovalWidget/PanelApprovalWidget.tsx [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+  if (submission.submissionState !== 'Approved') return null;
+
+  if (submission.createdBy.email === user.email) {
+    return (
+      <Banner
+        title="This submission needs approval"
+        className="lbh-page-announcement--info"
+      >
+        You cannot panel approve your own submission. Ask a manager or colleague
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApprovalWidget.tsx:29]: try

"You cannot approve your own submissions. Ask a manager or colleague for help"

File: components/ApprovalWidget/PanelApprovalWidget.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    );
+  }
+
+  return (
+    <>
+      <Banner
+        title="This submission needs approval"
+        className="lbh-page-announcement--info"
+      >
+        <p>Is this ready for panel approval?</p>
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApprovalWidget.tsx:41]: "Has the panel approved this work?"

File: components/ApprovalWidget/PanelApprovalWidget.tsx [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        <div className={s.actions}>
+          <button
+            className="lbh-link"
+            onClick={() => setApprovalDialogOpen(true)}
+          >
+            Yes, panel approves
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApprovalWidget.tsx:48]: "Yes, panel has approved"

**Comment by JohnFarrellDev** [comment]
It might be worth waiting for the API work to be on staging so we can fully check the functionality before merging.

We also need some tests.

File: components/ApprovalWidget/PanelApprovalWidget.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}: Props): React.ReactElement | null => {
+  const [approvalDialogOpen, setApprovalDialogOpen] = useState<boolean>(false);
+  const [rejectDialogOpen, setRejectDialogOpen] = useState<boolean>(false);
+
+  if (submission.submissionState !== 'Approved') return null;
+
+  if (submission.createdBy.email === user.email) {
+    return (
+      <Banner
+        title="This submission needs approval"
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApprovalWidget.tsx:26]: "This submission needs panel approval"

File: components/ApprovalWidget/PanelApprovalWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        You cannot panel approve your own submission. Ask a manager or colleague
+        for help.
+      </Banner>
+    );
+  }
+
+  return (
+    <>
+      <Banner
+        title="This submission needs approval"
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApprovalWidget.tsx:38]: "This submission needs panel approval"

File: components/ApprovalWidget/PanelApproveDialogue.tsx [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      setLoading(false);
+      setOpen(false);
+    } catch (e) {
+      setStatus(e.toString());
+    }
+  };
+
+  return (
+    <Dialog
+      title="Are you sure you want to approve this submission?"
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApproveDialogue.tsx:36]: "Are you sure you want to mark this submission as panel approved?"

File: components/ApprovalWidget/PanelApproveDialogue.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+            className="lbh-page-announcement--warning"
+          >
+            <p>Please refresh the page or try again later.</p>
+            <p className="lbh-body-xs">{status}</p>
+          </Banner>
+        )}
+        <p>Approved submissions can no longer be edited.</p>
+        <p>
+          If you&apos;re not sure what this means, check with a manager first.
+        </p>
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApproveDialogue.tsx:53]: replace with mandatory checkbox with "Yes, the panel met and approved this work"

File: components/ApprovalWidget/PanelApproveDialogue.tsx [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        <button
+          className="govuk-button lbh-button"
+          onClick={handleApprove}
+          disabled={loading}
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApproveDialogue.tsx:57]: add `useState` linked to the checkbox to disable this conditionally

File: components/ApprovalWidget/PanelApproveDialogue.tsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+        >
+          Yes, approve
```
Comment (jhackett1) [inline components/ApprovalWidget/PanelApproveDialogue.tsx:59]: Yes, mark as approved

**Comment by jhackett1** [comment]
should include "marked as panel approved by x" on the history timeline

File: lib/submissions.ts [hunk 10]
```diff
@@ -178,6 +178,23 @@ export const approveSubmission = async (
...
+/** mark an approved submission as panal approved passing in the id and approved by  */
+export const panelApproveSubmission = async (
+  submissionId: string,
+  approvedBy: string
+): Promise<Submission> => {
+  const { data } = await axios.patch(
+    `${ENDPOINT_API}/submissions/${submissionId}`,
+    {
+      editedBy: approvedBy,
+      submissionState: 'panelApproved',
```
Comment (JohnFarrellDev) [inline lib/submissions.ts:190]: Could the string value be `panel-approved` instead. It would be nicer for the backend.

Comment (anabebs) [reply lib/submissions.ts:190]: sure

**Comment by jhackett1** [comment]
Capture

**Comment by JohnFarrellDev** [comment]
I noticed `RejectDialog.tsx` exported component is called `ApproveDialog`, it's just a copy and paste error from before but should be named correctly

File: components/ApprovalWidget/PanelApproveDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+                  className="govuk-error-message lbh-error-message govuk-!-margin-top-2"
+                  role="alert"
+                >
+                  <span className="govuk-visually-hidden">Error:</span>
+                  {msg}
+                </p>
+              )}
+            </ErrorMessage>
+
+            <button className="govuk-button lbh-button" disabled={isSubmitting}>
```
Comment (JohnFarrellDev) [inline components/ApprovalWidget/PanelApproveDialog.tsx:87]: This should be disabled if the checkbox hasn't been ticked for "The panel met and approved this work" right?

Comment (JohnFarrellDev) [reply components/ApprovalWidget/PanelApproveDialog.tsx:87]: The button doesn't submit but visually would be nice to show it is disabled

**Review by JohnFarrellDev — APPROVED** [review]

File: components/ApprovalWidget/PanelApproveDialog.tsx (hunk 11 above)
Comment (jhackett1) [reply components/ApprovalWidget/PanelApproveDialog.tsx:87]: that's taken care of by the validation schema on the form that surrounds this

Comment (JohnFarrellDev) [reply components/ApprovalWidget/PanelApproveDialog.tsx:87]: I noticed when I went to the dialog the button wasn't disabled even though the checkbox had not been ticked yet. Just a visual bug.

**Pull request merged**